File: launcher/ClatterDrive.Launcher/BackendSettings.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace ClatterDrive.Launcher;

public sealed class BackendSettings
{
    public string Host { get; init; } = "127.0.0.1";
    public int Port { get; init; } = 8080;
    public string BackingDirectory { get; init; } = Path.GetFullPath("backing_storage");
    public string AudioMode { get; init; } = "live";
    public string? AudioDevice { get; init; }
    public string? AudioTeePath { get; init; }
    public string? EventTracePath { get; init; }
    public string DriveProfile { get; init; } = "desktop_7200_internal";
    public string? AcousticProfile { get; init; }
    public bool ColdStart { get; init; } = true;
    public bool AsyncPowerOn { get; init; } = true;

    public string WebDavUrl => $"http://{Host}:{Port}";

    public IReadOnlyList<string> ToServeArguments(bool jsonStatus = true)
    {
        var args = new List<string>
        {
            "serve",
            "--host",
            Host,
            "--port",
            Port.ToString(System.Globalization.CultureInfo.InvariantCulture),
            "--backing-dir",
            BackingDirectory,
            "--audio",
            AudioMode,
            "--drive-profile",
            DriveProfile,
        };
        if (jsonStatus)
        {
            args.Add("--json-status");
        }
        if (!string.IsNullOrWhiteSpace(AcousticProfile))
        {
            args.Add("--acoustic-profile");
            args.Add(AcousticProfile);
        }
        if (!string.IsNullOrWhiteSpace(AudioDevice))
        {
            args.Add("--audio-device");
            args.Add(AudioDevice);
        }
        if (!string.IsNullOrWhiteSpace(AudioTeePath))
        {
            args.Add("--audio-tee-path");
            args.Add(AudioTeePath);
        }
        if (!string.IsNullOrWhiteSpace(EventTracePath))
        {
            args.Add("--event-trace-path");
            args.Add(EventTracePath);
        }
        if (!ColdStart)
        {
            args.Add("--ready");
        }
        if (!AsyncPowerOn)
        {
            args.Add("--sync-power-on");
        }
        return args;
    }

    public Dictionary<string, string> ToEnvironment()
    {
        var env = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["FAKE_HDD_HOST"] = Host,
            ["FAKE_HDD_PORT"] = Port.ToString(System.Globalization.CultureInfo.InvariantCulture),
            ["FAKE_HDD_BACKING_DIR"] = BackingDirectory,
            ["FAKE_HDD_AUDIO"] = AudioMode,
            ["FAKE_HDD_DRIVE_PROFILE"] = DriveProfile,
            ["FAKE_HDD_COLD_START"] = ColdStart ? "on" : "off",
            ["FAKE_HDD_ASYNC_POWER_ON"] = AsyncPowerOn ? "on" : "off",
        };
        if (!string.IsNullOrWhiteSpace(AcousticProfile))
        {
            env["FAKE_HDD_ACOUSTIC_PROFILE"] = AcousticProfile!;
        }
        if (!string.IsNullOrWhiteSpace(AudioDevice))
        {
            env["FAKE_HDD_AUDIO_DEVICE"] = AudioDevice!;
        }
        if (!string.IsNullOrWhiteSpace(AudioTeePath))
        {
            env["FAKE_HDD_AUDIO_TEE_PATH"] = AudioTeePath!;
        }
        if (!string.IsNullOrWhiteSpace(EventTracePath))
        {
            env["FAKE_HDD_EVENT_TRACE_PATH"] = EventTracePath!;
        }
        return env;
    }
}

File: launcher/ClatterDrive.MacLauncher/Sources/ClatterDriveMacCore/BackendSettings.swift
import Foundation

public enum AudioMode: String, CaseIterable, Identifiable {
    case live
    case off

    public var id: String { rawValue }
}

public struct BackendSettings: Equatable {
    public var host: String
    public var port: Int
    public var backingDirectory: String
    public var audioMode: AudioMode
    public var audioDevice: String
    public var audioTeePath: String
    public var eventTracePath: String
    public var driveProfile: String
    public var acousticProfile: String
    public var coldStart: Bool
    public var asyncPowerOn: Bool

    public init(
        host: String = "127.0.0.1",
        port: Int = 8080,
        backingDirectory: String = FileManager.default.currentDirectoryPath + "/backing_storage",
        audioMode: AudioMode = .live,
        audioDevice: String = "",
        audioTeePath: String = "",
        eventTracePath: String = "",
        driveProfile: String = "seagate_ironwolf_pro_16tb",
        acousticProfile: String = "mounted_in_case",
        coldStart: Bool = true,
        asyncPowerOn: Bool = true
    ) {
        self.host = host
        self.port = port
        self.backingDirectory = backingDirectory
        self.audioMode = audioMode
        self.audioDevice = audioDevice
        self.audioTeePath = audioTeePath
        self.eventTracePath = eventTracePath
        self.driveProfile = driveProfile
        self.acousticProfile = acousticProfile
        self.coldStart = coldStart
        self.asyncPowerOn = asyncPowerOn
    }

    public var webDavURL: String {
        "http://\(host):\(port)"
    }

    public func validationMessage() -> String? {
        if backingDirectory.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return "Choose a backing folder."
        }
        if host.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return "Enter a host."
        }
        if port < 1 || port > 65_535 {
            return "Choose a port from 1 to 65535."
        }
        if driveProfile.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return "Choose a drive profile."
        }
        return nil
    }

    public func serveArguments(jsonStatus: Bool = true) -> [String] {
        var args = [
            "serve",
            "--host",
            host,
            "--port",
            String(port),
            "--backing-dir",
            backingDirectory,
            "--audio",
            audioMode.rawValue,
            "--drive-profile",
            driveProfile,
        ]
        if jsonStatus {
            args.append("--json-status")
        }
        if !acousticProfile.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            args.append(contentsOf: ["--acoustic-profile", acousticProfile])
        }
        if !audioDevice.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            args.append(contentsOf: ["--audio-device", audioDevice])
        }
        if !audioTeePath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            args.append(contentsOf: ["--audio-tee-path", audioTeePath])
        }
        if !eventTracePath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            args.append(contentsOf: ["--event-trace-path", eventTracePath])
        }
        if !coldStart {
            args.append("--ready")
        }
        if !asyncPowerOn {
            args.append("--sync-power-on")
        }
        return args
    }

    public func environment() -> [String: String] {
        var env = [
            "FAKE_HDD_HOST": host,
            "FAKE_HDD_PORT": String(port),
            "FAKE_HDD_BACKING_DIR": backingDirectory,
            "FAKE_HDD_AUDIO": audioMode.rawValue,
            "FAKE_HDD_DRIVE_PROFILE": driveProfile,
            "FAKE_HDD_COLD_START": coldStart ? "on" : "off",
            "FAKE_HDD_ASYNC_POWER_ON": asyncPowerOn ? "on" : "off",
        ]
        if !acousticProfile.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            env["FAKE_HDD_ACOUSTIC_PROFILE"] = acousticProfile
        }
        if !audioDevice.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            env["FAKE_HDD_AUDIO_DEVICE"] = audioDevice
        }
        if !audioTeePath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            env["FAKE_HDD_AUDIO_TEE_PATH"] = audioTeePath
        }
        if !eventTracePath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            env["FAKE_HDD_EVENT_TRACE_PATH"] = eventTracePath
        }
        return env
    }
}
